core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1969,34 +1954,54 @@ class KafkaApis(val requestChannel: RequestChannel,
-      sendResponse(request, createResponse(requestThrottleMs))
-    })
+  private def sendResponseMaybeThrottle(request: RequestChannel.Request,
+                                        createResponse: Int => AbstractResponse): Unit = {
+    sendResponseMaybeThrottle(request, throttleTimeMs => Some(createResponse(throttleTimeMs)),
+      closeIfNoResponse = false)
   }
 
-  private def sendResponseMaybeThrottle(request: RequestChannel.Request, clientId: String, sendResponseCallback: Int => Unit) {
-
+  private def sendResponseMaybeThrottle(request: RequestChannel.Request,
+                                        createResponse: Int => Option[AbstractResponse],
+                                        closeIfNoResponse: Boolean) {
     if (request.apiRemoteCompleteTimeNanos == -1) {
       // When this callback is triggered, the remote API call has completed
       request.apiRemoteCompleteTimeNanos = time.nanoseconds
     }
-    quotas.request.maybeRecordAndThrottle(request.session.sanitizedUser, clientId,
-        request.requestThreadTimeNanos, sendResponseCallback,
-        callback => request.recordNetworkThreadTimeCallback = Some(callback))
+
+    quotas.request.maybeRecordAndThrottle(request.session.sanitizedUser, request.header.clientId,
+      request.requestThreadTimeNanos,
+      sendResponseCallback = throttleTimeMs => sendResponse(request, createResponse(throttleTimeMs), closeIfNoResponse),
+      recordNetworkThreadTimeCallback = callback => request.recordNetworkThreadTimeCallback = Some(callback))
   }
 
-  private def sendResponseExemptThrottle(response: RequestChannel.Response) {
-    sendResponseExemptThrottle(response.request, () => requestChannel.sendResponse(response))
+  private def sendResponseExemptThrottle(request: RequestChannel.Request, response: AbstractResponse): Unit = {
+    sendResponseExemptThrottle(request, Some(response), closeIfNoResponse = false)
   }
 
-  private def sendResponseExemptThrottle(request: RequestChannel.Request, sendResponseCallback: () => Unit) {
+  private def sendResponseExemptThrottle(request: RequestChannel.Request,
+                                         response: Option[AbstractResponse],
+                                         closeIfNoResponse: Boolean): Unit = {
     quotas.request.maybeRecordExempt(request.requestThreadTimeNanos,
-        callback => request.recordNetworkThreadTimeCallback = Some(callback))
-    sendResponseCallback()
+      callback => request.recordNetworkThreadTimeCallback = Some(callback))
+    sendResponse(request, response, closeIfNoResponse)
   }
 
-  private def sendResponse(request: RequestChannel.Request, response: AbstractResponse) {
-    requestChannel.sendResponse(RequestChannel.Response(request, response))
+  private def sendResponse(request: RequestChannel.Request,
+                           responseOpt: Option[AbstractResponse],
+                           closeIfNoResponse: Boolean): Unit = {
```

Comment:
The goal was to have only one location where the `Response` object was built, but I agree the result turned out a bit goofy. Let me reconsider it. What we're really trying to do here is close the connection if we encounter an error and the client does not expect a response, so maybe let me see if I can refactor in those terms. Failing that, I'll just revert I guess.